### ISSUE TITLE
Refactor memorylimitprocessor to use built-in memory limiting in Go

### ISSUE THREAD

**Issue by tigrannajaryan** [open-telemetry/opentelemetry-collector#5708]
There is a [SetMemoryLimit](https://pkg.go.dev/runtime/debug@master#SetMemoryLimit) in Go runtime since 1.19. It is worth checking if using this is better than the manual limit that we do in the memorylimitprocessor currently. 

Also worth exploring if this can be possibly combined with GOGC=off (crazy?) or GOGC setting very high and deleting the ballast extension.

**Comment by tigrannajaryan** [comment]
> Unsetting GOGC is indeed crazy, as every workload is different, and every deployment has its memory usage profile. Containerized instances focused on distributed tracing have different requirements than bare-metal metrics-specific instances.

Indeed. And obviously it is totally wrong if the memorylimiting is not used (which is optional).

**Comment by mx-psi** [comment]
Any change that uses functions available only on Go 1.19 should be done under the `go1.19` build tag while keeping the existing implementation under `!go1.19`, to preserve the compatibility with Go 1.18 https://github.com/open-telemetry/opentelemetry-collector#compatibility

**Comment by seankhliao** [comment]
My reading is:

The memorylimiterprocessor is still useful as feedback into the pipeline to drop "real" memory usage down (if you prefer losing some data vs potentially crashing). It could be useful for it pick defaults based on the new memory limit settings if provided via the environment.

The ballast extension should probably be deprecated in favor of the memory limit. But it does have the advantage of reading from cgroup memory limits so maybe that could stay somewhere and set the gc memory limit.

For containerized environments with a guaranteed memory resource allocation, setting `GOGC=off` if there's a memory limit in effect sounds like a good tradeoff for some performance boost. For other environments I'm less sure what people expect.

**Comment by arielvalentin** [comment]
We have been trying to tune our collector gateway deployments to handle increased volume needs and find it quite challenging to do. Workloads in our busiest clusters generate about ~74K rps (~17.81M spans/sec) however we had to dedicate an unexpected amount of CPU and Memory to accommodate this subset of our total load. 

During our peak times we would see a spike in latency and the memory limiter start to kick in a subset of our collectors resulting in p99 request latency of ~10 sec:

![image](https://github.com/user-attachments/assets/38178203-4679-4ab6-ae35-39f5a957c68d)

We attempted to upgrade to v0.105.x and v0.106.x but something that we found perplexing though was that although we saw an increase in CPU and latency, we did not see an equivalent spike in memory so the memory limiter would not kick in. 

After several failed attempts to upgrade, we started profiling and found that there was an unexpected increase in in allocations due to internal telemetry and reported it to the core collector team[^2]. 

The profiler results also revealed that our OTTL statements were under optimized and largely contributing to the allocations and heap size growth. Since our request volume is spiky during the day, we would really only observe this being a problem under heavy load. 

We started experimenting with using `GOGC`[^1] and found that it has dramatically improved latency:

![image](https://github.com/user-attachments/assets/04e67233-7168-49cb-b66c-9fd5141d6bf8)

> For containerized environments with a guaranteed memory resource allocation, setting GOGC=off if there's a memory limit in effect sounds like a good tradeoff for some performance boost. For other environments I'm less sure what people expect.

Based on @seankhliao comment, it seems like we would be good candidates for disabling `GOGC` and then rely on the memory limiter, `GOMEMLIMIT`, and k8s OOM Killer to keep our workloads under control. 

Does this seem like a sensible thing to do?

[^1]: https://cloud-native.slack.com/archives/C01N6P7KR6W/p1723497498461129
[^2]: https://github.com/open-telemetry/opentelemetry-collector/issues/10858
